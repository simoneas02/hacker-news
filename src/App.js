import React, { Component } from 'react';

import List from './List';
import Search from './Search';

const DEFAULT_QUERY='redux';
const PATH_BASE='https://hn.algolia.com/api/v1';
const PATH_SEARCH='/search';
const PARAM_SEARCH='query=';

class App extends Component {
  constructor ( props ) {
    super(props);
    this.state = {
      title: 'My Title',
      result: null,
      searchTerm: DEFAULT_QUERY
    };

    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onDimiss = this.onDimiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
  }

  setSearchTopStories(result) {
    this.setState({ result });
  }

  onSearchSubmit() {
    const { searchTerm } = this.state; 
    this.fetchSearchTopStories(searchTerm);
  }

  fetchSearchTopStories(searchTerm) {
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}`)
      .then(response => response.json())
      .then(result => this.setSearchTopStories(result))
      .catch(error => error);
  }


  componentDidMount() {
    const { searchTerm } = this.state; 
    this.fetchSearchTopStories(searchTerm);
}

  onDimiss ( id ) {
    const isNotId = item => item.objectID !== id;
    const updateList = this.state.result.hits.filter( isNotId );
    this.setState({ result: { ...this.state.result, hits: updateList } });
  }

  onSearchChange =  event => {
    const { searchTerm } = this.state;
    this.fetchSearchTopStories(searchTerm);
    
  }

  render() {
    const { 
      title, 
      searchTerm,
      result 
    } = this.state;
    
    if (!result) { return null; }

    return (
      <div>
        <h1>{ title }</h1>

        <Search
          value={ searchTerm }
          onChange={ this.onSearchChange }
          onSearchSubmit={ this.onSearchSubmit }
        >
          Search
        </Search>

        { result &&
          <List
            list={ result.hits }
            pattern={ searchTerm }
            onDimiss={ this.onDimiss }
          />
        }
      </div>
    );
  }
}

export default App;
