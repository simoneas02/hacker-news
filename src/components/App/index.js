import React, { Component } from "react";
import axios from "axios";
import PropTypes from "prop-types";

import Button from "../Button";
import ListItem from "../ListItem";
import { Loading } from '../Loading';
import Search from "../Search";

import {
  DEFAULT_QUERY,
  DEFAULT_HPP,
  PATH_BASE,
  PATH_SEARCH,
  PARAM_SEARCH,
  PARAM_PAGE,
  PARAM_HPP
} from "../../constants";

class App extends Component {
  _isMounted = false;

  constructor(props) {
    super(props);
    this.state = {
      title: "My Title",
      results: null,
      searchTerm: DEFAULT_QUERY,
      searchKey: " ",
      error: null,
      isLoading: false
    };

    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onDimiss = this.onDimiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
  }

  fetchSearchTopStories(searchTerm, page = 0) {
    this.setState({ isLoading: true });
    axios
      .get(
        `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`
      )
      .then(result => this._isMounted && this.setSearchTopStories(result.data))
      .catch(error => this._isMounted && this.setState({ error }));
  }

  componentDidMount() {
    this._isMounted = true;
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  needsToSearchTopStories(searchTerm) {
    return !this.state.results[searchTerm];
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });

    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm);
    }

    event.preventDefault();
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value });
  }

  setSearchTopStories(result) {
    const { hits, page } = result;
    const { searchKey, results } = this.state;
    const oldHits =
      results && results[searchKey] ? results[searchKey].hits : [];

    const updatedHits = [...oldHits, ...hits];
    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      },
      isLoading: false
    });
  }

  onDimiss(id) {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const isNotId = item => item.objectID !== id;
    const updatedHits = hits.filter(isNotId);

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      }
    });
  }

  render() {
    const { title, searchTerm, results, searchKey, error, isLoading } = this.state;

    const page =
      (results && results[searchKey] && results[searchKey].page) || 0;

    const list =
      (results && results[searchKey] && results[searchKey].hits) || [];

    return (
      <div>
        <h1>{title}</h1>

        <Search
          value={searchTerm}
          onChange={this.onSearchChange}
          onSubmit={this.onSearchSubmit}
        >
          Search
        </Search>

        {error ? (
          <p>Something went wrong.</p>
        ) : (
          results && (
            <ul>
              <ListItem list={list} onDimiss={this.onDimiss} />
            </ul>
          )
        )}

        <div className="interactions">
          {isLoading
            ? <Loading />
            : <Button
                onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}
              >
              More
            </Button>
          }
        </div>
      </div>
    );
  }
}

App.propTypes = {
  title: PropTypes.string,
  results: PropTypes.array,
  searchTerm: PropTypes.string,
  searchKey: PropTypes.string,
  error: PropTypes.element,
  isLoading: PropTypes.bool
};

export default App;
