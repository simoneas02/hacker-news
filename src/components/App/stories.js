import React from 'react';
import { storiesOf } from '@storybook/react';

storiesOf('App', module).add("I'm App", () => <h1>Hello App</h1>);
