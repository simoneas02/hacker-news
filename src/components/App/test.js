import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import App from '.';

Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {
  const props = {
    title: 'My Title',
    results: null,
    searchTerm: 'redux',
    searchKey: ' ',
    error: null
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('rendered the title', () => {
    const element = shallow(<App />);
    expect(element.state().title).toContain('My Title');
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
