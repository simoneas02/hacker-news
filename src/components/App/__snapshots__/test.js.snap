// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div>
  <h1>
    My Title
  </h1>
  <form
    onSubmit={[Function]}
  >
    Search
    <input
      onChange={[Function]}
      type="text"
      value="redux"
    />
    <button
      type="submit"
    >
      Search
    </button>
  </form>
  <div
    className="interactions"
  >
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      More
    </button>
  </div>
</div>
`;
